### ISSUE TITLE
Shift... Shift everywhere...

### ISSUE THREAD

**Issue by Svarov-RZM** [synfig/synfig#3472]
### Description
![ShiftEverywhere](https://github.com/user-attachments/assets/d44c1a8d-6320-4580-8dec-90aace3586a5)

Okay, there's a serious problem in Synfig where a considerable amount of layers has a slight shift in them (like 1-2 px) unlike versions <= 1.2.2 that don't. Shift is not only present in the gradients/distortions but even in some composite layers like "Shade" layer for example.

### To Reproduce
Here's a test-case file with this layer:
[ShadeTest.zip](https://github.com/user-attachments/files/18220947/ShadeTest.zip)

Unpack, switch in image viewer between rendered pngs and you will notice that version 1.4.4 (and newer) has a slight shift toward bottom-right corner.

### Additional
This problem starts to drive me crazy. I wanted to test if I can use "Shade" layer to make linework based on a complex shape and I can't because of that damned shift that messes up the result. 

In this thread I discovered that the problem was introduced after 1.2.2 release: https://github.com/synfig/synfig/pull/3306 
Please read it if you haven't.

I am going to head-dive in the commits trying to find the culprit.
I'll report it here if I do and maybe someone can hopefully fix this annoying issue.

**Comment by rodolforg** [comment]
I accidently found this code slice:

https://github.com/synfig/synfig/blob/9f576ad5b30a427fcd901c4ebe00f0c15a89c636/synfig-core/src/synfig/target_scanline.cpp#L101-L109

If I comment it out, the shift doesn't happen.
But why is this a problem to some top layers only?

**Comment by Svarov-RZM** [comment]
Alright, just like I said I tried to build various versions of Synfig to see where exactly the error was introduced and failed :( I was able to go back to version < 1.3.11, exact commit: 2c0523022e2c5d828ca01726fd394d3fdb706b1d and the shift was already there.
I couldn't go past that because Synfig would no longer build on my Linux box, perhaps I needed the older version of Debian... Anyway, judging by what rodolforg said and the commit I stopped at, the shift was probably introduced somewhere along these two commits: da43984f62e7c453e1caab279e5bb11517883d90 and 5e27542ae28cd7fde5ec4995d8b7ca53a03daeb3

> But why is this a problem to some top layers only?

Hmm... I did some additional testing and found out something curios - the shift happens on layers that take **underlying context**, such as: shade, blur, distortions. Layers that draw themselves such as circle, are fine. A notable exception from this is "**Filter group**" - this layer also takes underlying context but doesn't introduce the shift. This leads me to believe that perhaps the renderer is fine but the problem is indeed in ports?

@rodolforg Can you check my theory? What is a difference between "Filter group" taking underlying context and other layers that's been ported to new engine, such as Shade, Blur?

**Comment by rodolforg** [comment]
Wow.
Lots of changes between those two commits.
Difficult to track :(

Hmm. The Plant layer also suffers with the pixel shift.

**Comment by rodolforg** [comment]
Wow.
Lots of changes between those two commits.
Difficult to track :(

Hmm. The Plant layer also suffers with the pixel shift.

The shift already exists in current dev releases, right?

**Comment by Svarov-RZM** [comment]
> Hmm. The Plant layer also suffers with the pixel shift.

Yeah, you are correct. There's also Noise Gradient that has shift but doesn't depend on underlying context. So my idea is wrong, but still, I think it worth checking the difference in rendering of Shade/Blur against Filter group because one has a problem and the other don't. I am not good with C++ but I'll check it, maybe I will find some clues...

> The shift already exists in current dev releases, right?

Right.

**Comment by Svarov-RZM** [comment]
Just for information, I was debugging this problem and discovered that indeed, resolution of problematic layers get changed before rendering, just as I stated here: https://github.com/synfig/synfig/pull/3306#issuecomment-2574930847

Let's take "Noise Gradient" layer as an example. If I understand correctly, the layer is drawn by TaskPaintPixelSW->run_task procedure (located in taskpaintpixelsw.cpp). There in the beginning I placed some debugging printfs that show the target resolution:

        ...
        if (!task->is_valid())
                return true;

        printf("[TaskPaintPixelSW->run_task]\n");
        printf(" [SOURCE] %f %f || %f %f\n", task->source_rect.minx, task->source_rect.miny, task->source_rect.maxx, task->source_rect.maxy);
        printf(" [TARGET] %i %i || %i %i\n", task->target_rect.minx, task->target_rect.miny, task->target_rect.maxx, task->target_rect.maxy);

        Vector ppu = task->get_pixels_per_unit();
        ...


If we grab TestNoiseGr3.zip from linked issue, set "Noise Gradient" layer's blend method to composite and render it at 270x270 px we get this output in console:

    [TaskPaintPixelSW->run_task]
     [SOURCE] -2,316667 -2,316667 || 2,316667 2,316667
     [TARGET] 0 0 || 274 274

The resolution of the target_rect has changed to 274x274 px! So it increased by 4 pixels and it's not proportional. If you render it at 500x500 px, the target_rec will become 504x504 px.

If we comment out the code snippet that was suggested by @rodolforg in target_scanline.cpp and try to render with the same settings, we'll get this:

    [TaskPaintPixelSW->run_task]
     [SOURCE] -2,250000 -2,250000 || 2,250000 2,250000
     [TARGET] 0 0 || 270 270

Notice that not only target_rect changed but also source_rect did and now both resolutions match what we specified in rendering dialogue. The reason why we can't keep those lines of code removed is because they are responsible for converting the order of pixels from Synfig to png. Without it, the png files would render upside down.

So basically, somewhere in the code before TaskPaintPixelSW->run_task, coordinates gets screwed up but only for certain layers that use TaskPaintPixelSW and some other tasks (e.g. TaskPixelColorMatrixSW). Hopefully, what I found will give someone a clue where else to look for a potential error. I'll get back if I find anything else.

**Comment by ice0** [comment]
Wow! That makes sense, but I need to think over what you wrote.  
Generally yes, there was an assumption that the shift is caused by resampling, but I didn’t have time to debug it.  
In any case, I suggest creating a Pull Request, as it will be easier to understand the changes with the code.

**Comment by Svarov-RZM** [comment]
Hey guys,

I've been debugging this goddamn pixel shift for quite a while and I have finally some progress to share. It seems that the issue is in the wrong matrix transformation in synfig-core\src\synfig\rendering\software\task\tasktransformationaffinesw.cpp. Look at this code:

<img width="915" height="416" alt="Image" src="https://github.com/user-attachments/assets/8fa35429-1294-4646-80a6-1fffae03ed68" />

You can see I put a debug printf there which will give the following line when rendering one frame from the example above in this thread:

`Matrix: 1,000000 -1,000000 -4,000000 644,000000`

The resolution of the example file is 640x640, 644,000000 in the output comes from blur, it extends bounding rectangle, so it seems fine. But let's experiment with these values a little.

From various attempts, I found out that adjusting the matrix by -0.5 fixes the shift. I don't know why, I am very poor with C++ and algorithms, so it's all comes from trial and error. Anyway, the resulting change would be:

<img width="914" height="459" alt="Image" src="https://github.com/user-attachments/assets/55641261-a207-46ba-99c9-044986e6c226" />

And that fixes the 1px shift! I tested it on everything I could, including the example from here: https://github.com/synfig/synfig/pull/3530
All works properly with this change, but again, I do not understand the root case, why it's necessary to adjust matrix by 0.5... Why this value and not other? Maybe some of you with better C++ skills understands what's going on? Should I make a PR? I already made a mistake trying to fix it if you remember...

@rodolforg @ice0 @BobSynfig @Abdelhadi-Wael

**Comment by Abdelhadi-Wael** [comment]
Hi @Svarov-RZM

I spent some time debugging this and here is what I figured out:


### **What happens**

1. `tasktransformationaffinesw.cpp` returns coordinates in the integer corners system.
2. It passes these to `resample.cpp`, which is subtracting .5 (not sure why maybe was expecting centers? ) :
```cpp
i.pos = back_transformation.get_transformed( start ) - Vector(0.5, 0.5);

```
Because the input was a corner (integer), subtracting 0.5 lands the sampler on a .5 boundary

3. `synfig-core/src/synfig/surface_etl.h` receives that coordinate If the coordinate is not an integer (like the .5 boundary we just created), it rounds it to an integer value:
```cpp
/** Retrieve the nearest sample */
static value_type nearest_sample(const void *surface, const coord_type x, const coord_type y)
    { return (value_type)reader(surface, synfig::round_to_int(x), synfig::round_to_int(y)); }

```

now lets see how the problem happens exactly with an example to clarify:

We want to sample the color at **Pixel (4, 595)**

<img width="621" height="341" alt="Image" src="https://github.com/user-attachments/assets/a277a93f-a4ac-49d5-8415-fa94bc96dc09" />

### 1. on X-Axis 


A. current Code 

1. Matrix returns the Corner: 4.0
2. Resampler subtracts 0.5: 3.5
3. Sampler Calculation: round_to_int(3.5) -> 4 (correct)

B. Shift the matrix (get centers instead of corners)

1. Matrix Shift (-0.5) returns the center: 4.5
2. Resampler subtracts 0.5: 4.0
3. Sampler Calculation: round_to_int(4) -> 4 (correct)



### 2. on Y-Axis (Flipped)


A. Current Code: 

1. Matrix returns the Corner : 596.0 (because it is flipped)
2. Resampler subtracts 0.5: 595.5
3. Sampler Calculation: round_to_int(595.5) -> 596 (wrong this is the cause of the shift)


B. Shift the matrix (get centers instead of corners)

1. Matrix Shift (-0.5) returns the center : 595.5
2. Resampler subtracts 0.5:  595.0
3. Sampler Calculation: round_to_int(595) -> 595 (correct)

**Comment by rodolforg** [comment]
Great work, @Svarov-RZM and @Abdelhadi-Wael ! :)

So, maybe the solution would be something like this in the same snippet of tasktransformationaffinesw.cpp that Svarov mentioned?

<img width="942" height="169" alt="Image" src="https://github.com/user-attachments/assets/75534a69-e69a-459d-8e60-3db84ad2c336" />

I'm not sure about it yet. Rotations would cause the same issue? This code change would make it correct or would introduce a (new?) error?

**Comment by Svarov-RZM** [comment]
> 2\. It passes these to `resample.cpp`, which is subtracting .5 (not sure why maybe was expecting centers? ) :

Yeah, I don't get it either. I tried to remove those 0.5 corrections in resample.cpp but it gave no visual difference in rendering.

> 3\. `synfig-core/src/synfig/surface_etl.h` receives that coordinate If the coordinate is not an integer (like the .5 boundary we just created), it rounds it to an integer value:

Oh, so this is the procedure that performs rounding, I missed it, didn't dig deep enough...

> I'm not sure about it yet. Rotations would cause the same issue? This code change would make it correct or would introduce a (new?) error?

Hmm, I've done some tests and I am pretty sure we can safely subtract 0.5 in all cases, no if condition is required.

Okay, let's try to summarize everything we know and try to make a test with different variables to see if our change in matrix might introduce other regressions.

First, why do we need to flip the image before rendering? It's because the target format store pixels in a different way than we do. Png format stores pixels in top-to-bottom order, so we have to insert transformation into task query. Bmp format, on the other hand, stores pixels in the same way we do (as I understand) - bottom-to-top so no additional transformation is required. If you render TwoCircles.sif (attached above in this thread) as bmp target you will see there's no shift. When we need to flip the image though, we get the shift during to a rounding error pointed by @Abdelhadi-Wael.

Now, I made this example to test different cases:

<img width="480" height="270" alt="Image" src="https://github.com/user-attachments/assets/7a05c80f-5874-49f2-94b1-d4dfdf665bcc" />

Source:
[StarTest.zip](https://github.com/user-attachments/files/24477831/StarTest.zip)

It has a bunch of cases in it, from top-left to bottom-right: 
* 2 Star layers, one colored black with a very small amount of blur to trigger resampling
* 1 Star layer and a Shade layer with blur
* 3 Star layers all with different amount of blur
* 1 Star layer with blur, rotated
* 1 Star layer with blur, rotated and scaled
* 1 Star layer with blur, rotated, scaled and skewed

Let's check the difference in rendering between 1.0.2 and 1.5.3 versions:

<img width="480" height="270" alt="Image" src="https://github.com/user-attachments/assets/1313a3ce-98fb-4930-8e29-567fd2ed0034" />

The image you're seeing is generated by ImageMagick, I subtracted two images and then performed leveling of colors to highlight the difference. You can see that all 6 examples from the file aren't clear stars - that's because of the offset in 1.5.3 version.

Now we compare 1.0.2 against Master with above matrix corrections:

<img width="480" height="270" alt="Image" src="https://github.com/user-attachments/assets/def83069-684b-45f7-9d5b-83c3181a9508" />

The star shapes are much more pronounced now because offset is gone. You might wonder why there's a difference in the first place, shouldn't the resulting image be solid black? No, because there was a rewrite in 1.4.x version that changed how blur behaves and so after subtracting images there's still a thin line left because of the blur that gets amplified by color leveling operation.

Rendering this example to bmp target also produce proper results, so I think no other regression were introduced, hopefully. I'll wait a couple of days and then make a PR unless you guys think the proposed solution is wrong.
